keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlVerificationRepository.kt

```diff
@@ -1,25 +1,122 @@
 import org.jooq.DSLContext
+import org.jooq.Field
 import org.jooq.Record1
 import org.jooq.ResultQuery
 import org.jooq.Select
+import org.jooq.impl.DSL.isnull
+import org.jooq.impl.DSL.select
 import java.time.Clock
+import java.time.Duration
+import java.time.Instant.EPOCH
+import java.time.LocalDateTime
+import java.time.ZoneOffset.UTC
 
 class SqlVerificationRepository(
-  private val jooq: DSLContext,
-  private val clock: Clock
-) : VerificationRepository {
+  jooq: DSLContext,
+  clock: Clock,
+  resourceSpecIdentifier: ResourceSpecIdentifier,
+  objectMapper: ObjectMapper,
+  sqlRetry: SqlRetry,
+  artifactSuppliers: List<ArtifactSupplier<*, *>> = emptyList(),
+  specMigrators: List<SpecMigrator<*, *>> = emptyList()
+) : SqlStorageContext(
+  jooq,
+  clock,
+  sqlRetry,
+  objectMapper,
+  resourceSpecIdentifier,
+  artifactSuppliers,
+  specMigrators
+), VerificationRepository {
+
+  override fun nextEnvironmentsForVerification(
+    minTimeSinceLastCheck: Duration,
+    limit: Int
+  ): Collection<VerificationContext> {
+    val now = clock.instant()
+    val cutoff = now.minus(minTimeSinceLastCheck).toTimestamp()
+    return sqlRetry.withRetry(WRITE) {
+      jooq.inTransaction {
+        select(
+          DELIVERY_CONFIG.UID,
+          DELIVERY_CONFIG.NAME,
+          ENVIRONMENT.UID,
+          ENVIRONMENT.NAME,
+          DELIVERY_ARTIFACT.UID,
+          DELIVERY_ARTIFACT.REFERENCE,
+          ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION
+        )
+          .from(ENVIRONMENT)
+          // join delivery config
+          .join(DELIVERY_CONFIG)
+          .on(DELIVERY_CONFIG.UID.eq(ENVIRONMENT.DELIVERY_CONFIG_UID))
+          // the application is not paused
+          .andNotExists(
+            selectOne()
+              .from(PAUSED)
+              .where(PAUSED.NAME.eq(DELIVERY_CONFIG.APPLICATION))
+              .and(PAUSED.SCOPE.eq(APPLICATION.name))
+          )
+          // join currently deployed artifact version
+          .join(ENVIRONMENT_ARTIFACT_VERSIONS)
+          .on(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+          .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(CURRENT.name))
+          // join artifact
+          .join(DELIVERY_ARTIFACT)
+          .on(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID))
+          // left join so we get results even if there is no row in ENVIRONMENT_LAST_VERIFIED
+          .leftJoin(ENVIRONMENT_LAST_VERIFIED)
+          .on(ENVIRONMENT_LAST_VERIFIED.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+          .and(ENVIRONMENT_LAST_VERIFIED.ARTIFACT_UID.eq(DELIVERY_ARTIFACT.UID))
+          // has not been checked recently (or has never been checked)
+          .where(ENVIRONMENT_LAST_VERIFIED.AT.orEpoch().lessOrEqual(cutoff))
+          // order by last time checked with things never checked coming first
+          .orderBy(ENVIRONMENT_LAST_VERIFIED.AT.orEpoch())
```

Comment:
I'm having trouble wrapping my head around this `orderBy` call. What is the SQL that it translates to?

Reply:
```order by ifnull(environment_last_verified.at, EPOCH)```

so basically if the left join to `environment_last_verified` did not find a row (hence `at` will be null) it uses epoch which will sort first.